core/src/main/java/org/apache/iceberg/SchemaUpdate.java

```diff
@@ -371,4 +455,54 @@ public Type primitive(Type.PrimitiveType primitive) {
     newFields.addAll(adds);
     return Types.StructType.of(newFields);
   }
+
+  private static class CompanionSchemaNewFieldsCollector extends TypeUtil.CompanionSchemaVisitor<TypeUtil.WrappedType> {
```

Comment:
I like the approach of using a companion visitor to collect the new fields, but I think it is a little too complex to use ownership. I'd prefer an approach that looks like this:

```
public T primitive(Type.PrimitiveType left, Type right);
```

In that case, left or right might be null and you can compare the two. Left might be an int and right might be a long, for example. That way, you don't have a case where ownership is both.